### DISCUSSION TITLE
Add action to crop/trim a sprite object

### DISCUSSION THREAD

**Discussion by tristanbob** [4ian/GDevelop#2843] [Enhancement ideas (and their technical discussions)]
## Description

I want the ability to cut out sections of a sprite that can then be used for programmatic destruction animations. 

For instance, to slice a sprite in half:

- Delete original sprite (full image)
- Create two new instances of that sprite, but crop opposite sides on each one
- Apply forces to each sprite for the animation

https://user-images.githubusercontent.com/8879811/104852730-db843200-58b9-11eb-83a6-8f54d01d511e.mp4

I also want to use this feature for other animations, such as exploding or crumbling.  I can fake this right now by using a full copy of the sprite:

https://user-images.githubusercontent.com/8879811/104852599-2a7d9780-58b9-11eb-86a8-d535d0e2722d.mp4

You can currently adjust the width and height of a sprite, but that simply does the same thing as changing scale; it stretches and compresses the sprite.

## Solution suggested

- [ ] Create an action for sprite objects that allows the user to select which region of the sprite stays and the rest of the sprite is removed.  

Inputs: 
- Sprite (object)
- Top-Left corner of the cropped region (x,y) relative to object origin
- Bottom-Right corner of the cropped region (x,y) relative to object origin

Output:
- Sprite object (only including the section inside the cropped region)

**Comment by ssangervasi** [comment]
A rectangle-only crop feature would meet your needs, but I'm curious if it would be generally useful. A rectangular crop is straightforward to do as a separate animation, and it wouldn't incur the runtime cost of masking the sprite.

However, something that would meet your need and have other uses would be the ability to have a [collision mask](http://wiki.compilgames.net/doku.php/gdevelop5/objects/sprite/collision-mask#collision_masks_for_sprite_objects) also apply as a visual mask. That would take advantage of the existing mask editor and you'd still get to reuse the sprite image.

**Comment by tristanbob** [comment]
@ssangervasi thanks for taking the time to respond.  My main goal with this feature is to write extensions that cut up a sprite into dozens of pieces that can then fall to the ground (or be exploded outward, or in a direction).  I don't understand how your suggestion helps me accomplish that.

**Comment by Bouh** [comment]
The way you do it in GDevelop seems to me to be the best possible solution.

Another solution would be to add prefabs to the game engine.
This subject is being discussed.

**Comment by tristanbob** [comment]
> A rectangle-only crop feature would meet your needs, but I'm curious if it would be generally useful. A rectangular crop is straightforward to do as a separate animation, and it wouldn't incur the runtime cost of masking the sprite.
> 
> However, something that would meet your need and have other uses would be the ability to have a [collision mask](http://wiki.compilgames.net/doku.php/gdevelop5/objects/sprite/collision-mask#collision_masks_for_sprite_objects) also apply as a visual mask. That would take advantage of the existing mask editor and you'd still get to reuse the sprite image.

@ssangervasi I think I finally understand what you mean; only show the part of the sprite that inside the standard collision hitbox.  That sounds like it could work, but I would need the ability to change that hitbox for each instance of an object, and do it using actions.

**Comment by ssangervasi** [comment]
> change that hitbox for each instance of an object, and do it using actions.

Yeah, I was thinking you'd have N animations defined with different hitboxes through the sprite properties, then use actions to switch animations
